### PR TITLE
Add metrics debugging cardinality APIs

### PR DESCRIPTION
**Pull request by AndrewHess**
# Description
1. Adds handlers for getting cardinality info about metrics (see the changes to API_README.md)
2. Fixes a bug with the tags tree reader; when there were more than `(1 << 16) / 8` TSIDs, we were getting overflow
3. Rearrange some code (like moving `parseAlphaNumTime()` into utils)
4. Add a new type `Epoch` which can be parsed as JSON and either be a unix epoch in seconds or be a string like `"now-1h"` that we want to convert to a unix epoch

# Testing
Manually tested calling the new endpoints

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [x] I have self-reviewed this PR.
- [x] I have removed all print-debugging and commented-out code that should not be merged.
- [x] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [x] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

File: pkg/segment/reader/metrics/tagstree/tagstreereader.go [hunk 1]
```diff
@@ -316,9 +351,11 @@ func (ttr *TagTreeReader) GetMatchingTSIDs(mName uint64, tagValue uint64, tagOpe
 		startOff = utils.BytesToUint32LittleEndian(ttr.metadataBuf[id : id+4])
 		id += 4
 		endOff = utils.BytesToUint32LittleEndian(ttr.metadataBuf[id : id+4])
+
 		tagTreeBuf := make([]byte, endOff-startOff)
 		_, err := ttr.fd.ReadAt(tagTreeBuf, int64(startOff))
 		if err != nil {
+			log.Errorf("TagTreeReader.GetMatchingTSIDs: failed to read tagtree buffer with startOffset %v and endOffset %v; err=%+v", startOff, endOff, err)
```
Comment (nkunal) [inline pkg/segment/reader/metrics/tagstree/tagstreereader.go:358]: nit: log the tagkey or the filename here

**Review by nkunal — DISMISSED** [review]
Fantastic PR! Nice and clean code.
The current approach of using maps to find uniqueness is fine. However as we scale the cardinalities we could run into mem issues if we storing that many uniqiness things.
In the next PR, we should look at either using HLL or bloom to store the unique tsdis, etc..

File: pkg/integrations/prometheus/promql/metricsSearchHandler.go [hunk 2]
```diff
@@ -882,6 +885,301 @@ func parseMetricTimeSeriesRequest(rawJSON []byte) (uint32, uint32, []map[string]
...
+		utils.SendInternalError(ctx, "Failed to search metrics", "Failed to get tags trees", err)
+		return
+	}
+
+	tagKeys := make(map[string]struct{})
+	for _, segmentTagTreeReader := range tagsTreeReaders {
+		tagKeys = utils.MergeMaps(tagKeys, segmentTagTreeReader.GetAllTagKeys())
+	}
+
+	allTsids := make(map[uint64]struct{})
```
[the remark this answers is not in the document]
Comment (AndrewHess) [reply pkg/integrations/prometheus/promql/metricsSearchHandler.go:923]: Good point, we should use HLL or a bloom instead. I'll do that in a later PR

File: pkg/integrations/prometheus/promql/metricsSearchHandler.go [hunk 3]
```diff
@@ -882,6 +885,301 @@ func parseMetricTimeSeriesRequest(rawJSON []byte) (uint32, uint32, []map[string]
...
+	for _, segmentTagTreeReader := range tagsTreeReaders {
+		for tagKey := range tagKeys {
+			tsids, err := segmentTagTreeReader.GetTSIDsForKey(tagKey)
```
[the remark this answers is not in the document]
Comment (AndrewHess) [reply pkg/integrations/prometheus/promql/metricsSearchHandler.go:926]: Good catch, I was erroring out in that case but just updated the PR with the fix. I manually tested by using sigclient to ingest two different sets of metrics into two different segments and then ran
```
curl -X POST -d '{"startEpoch":3881, "endEpoch":"now"}' \
http://localhost:5122/metrics-explorer/api/v1/series-cardinality
```
Now it's working as expected

**Pull request merged**